From: Hendrik Brueckner <brueckner@linux.ibm.com>
Subject: s390/cpum_sf: ensure sample frequency of perf event attributes is non-zero
Patch-mainline: v4.17-rc6
Git-commit: 4bbaf2584b86b0772413edeac22ff448f36351b1
References: bnc#1096746, LTC#168035

Description:  cpum_sf: ensure sample freq is non-zero
Symptom:      A kernel crash might occur.
Problem:      An FP divide exception during the sample rate
              initialization for the hardware sampling facility.
Solution:     Ensure that the sample frequency of the perf
              event attribute is non-zero.
Reproduction: Run trinity against the perf_event_open() syscall.

Upstream-Description:

              s390/cpum_sf: ensure sample frequency of perf event attributes is non-zero

              Correct a trinity finding for the perf_event_open() system call with
              a perf event attribute structure that uses a frequency but has the
              sampling frequency set to zero.  This causes a FP divide exception during
              the sample rate initialization for the hardware sampling facility.

              Fixes: 8c069ff4bd606 ("s390/perf: add support for the CPU-Measurement Sampling Facility")
              Cc: stable@vger.kernel.org # 3.14+
              Reviewed-by: Heiko Carstens <heiko.carstens@de.ibm.com>
              Signed-off-by: Hendrik Brueckner <brueckner@linux.ibm.com>
              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Hendrik Brueckner <brueckner@linux.ibm.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 arch/s390/kernel/perf_cpum_sf.c |    4 ++++
 1 file changed, 4 insertions(+)

--- a/arch/s390/kernel/perf_cpum_sf.c
+++ b/arch/s390/kernel/perf_cpum_sf.c
@@ -743,6 +743,10 @@ static int __hw_perf_event_init(struct p
 	 */
 	rate = 0;
 	if (attr->freq) {
+		if (!attr->sample_freq) {
+			err = -EINVAL;
+			goto out;
+		}
 		rate = freq_to_sample_rate(&si, attr->sample_freq);
 		rate = hw_limit_rate(&si, rate);
 		attr->freq = 0;
